### PR TITLE
[SPARK-42418][DOCS][PYTHON] PySpark documentation updates to improve discoverability and add more guidance

### PR DESCRIPTION
**Pull request by aldogrut**
### What changes were proposed in this pull request?
Updates to the PySpark documentation web pages that help users choose which API to use when and make it easier to discover relevant content and navigate the documentation pages. This is the first of a series of updates.

### Why are the changes needed?
The PySpark documentation web site does not do enough to help users choose which API to use and it is not easy to navigate.

### Does this PR introduce _any_ user-facing change?
Yes, the user facing PySpark documentation is updated.

### How was this patch tested?
Built and tested the PySpark documentation web site locally.

### CODE REVIEW

**Comment by zhengruifeng** [comment]
@allanf-db Thank you for this contribution!
Would you mind attaching the screenshot of updated document? so others can better understand the changes

**Comment by allanf-db** [comment]
Attaching screenshot of the updated document (Overview page).

The current page is here:
https://spark.apache.org/docs/latest/api/python/

NewOverviewPageTop

NewOverviewPageBottom

**Review by zhengruifeng — APPROVED** [review]

File: python/docs/source/index.rst [hunk 1]
```diff
@@ -17,58 +17,148 @@
...
-|binder|_ | `GitHub <https://github.com/apache/spark>`_ | `Issues <https://issues.apache.org/jira/projects/SPARK/issues>`_ | |examples|_ | `Community <https://spark.apache.org/community.html>`_
-
-PySpark is an interface for Apache Spark in Python. It not only allows you to write
-Spark applications using Python APIs, but also provides the PySpark shell for
-interactively analyzing your data in a distributed environment. PySpark supports most
-of Spark's features such as Spark SQL, DataFrame, Streaming, MLlib
-(Machine Learning) and Spark Core.
+**Date**: |today| **Version**: |release|
 
-.. image:: ../../../docs/img/pyspark-components.png
```
Comment (HyukjinKwon) [inline python/docs/source/index.rst:32]: I think we can remove `pyspark-components.png` in this repo.

File: python/docs/source/conf.py [hunk 2]
```diff
@@ -121,9 +121,12 @@
 # built documents.
 #
 # The short X.Y version.
-version = 'master'
+import pyspark
+import re
+version = re.sub(r"(\.?)dev.*", "", pyspark.__version__)
+
 # The full version, including alpha/beta/rc tags.
-release = os.environ.get('RELEASE_VERSION', version)
```
Comment (HyukjinKwon) [inline python/docs/source/conf.py:126]: hm, probably we should revert these changes so we can tell it's built from development version etc. Maybe for other vendors too (?).

**Review by HyukjinKwon — APPROVED** [review]
This is nice. LGTM otherwise

File: python/docs/source/index.rst (hunk 1 above)
Comment (allanf-db) [reply python/docs/source/index.rst:32]: I'll remove it

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Comment by itholic** [comment]
Late LGTM. It's really cool.
